File: clients/client/src/download.js
import got from 'got';
import util from 'util';
import stream from 'stream';
const pipeline = util.promisify(stream.pipeline);
import retry from './retry.js';
import { HashStream, ACCEPTABLE_HASHES } from './hashstream.js';
import taskcluster from './index.js';

// apply default retry config
const makeRetryCfg = ({ retries, delayFactor, randomizationFactor, maxDelay }) => ({
  retries: retries === undefined ? 5 : retries,
  delayFactor: delayFactor === undefined ? 100 : delayFactor,
  randomizationFactor: randomizationFactor === undefined ? randomizationFactor : 0.25,
  maxDelay: maxDelay === undefined ? 30 * 1000 : maxDelay,
});

const s3 = async ({ url, streamFactory, retryCfg }) => {
  return await retry(retryCfg, async (retriableError, attempt) => {
    let contentType = 'application/binary';
    try {
      const src = got.stream(url, { retry: false });
      src.on('response', res => {
        contentType = res.headers['content-type'] || contentType;
      });
      const dest = await streamFactory();
      await pipeline(src, dest);
      return contentType;
    } catch (err) {
      // treat non-500 HTTP responses as fatal errors, and retry everything else
      if (err instanceof got.HTTPError && err.response.statusCode < 500) {
        throw err;
      }
      return retriableError(err);
    }
  });
};

const getUrl = async ({ object, name, resp, streamFactory, retryCfg }) => {
  let responseUsed = false;
  let hashStream;
  let contentType = 'application/binary';

  await retry(retryCfg, async (retriableError, attempt) => {
    // renew the download URL if necessary (note that we assume the object-sevice
    // credentials are good for long enough)
    if (responseUsed && new Date(resp.expires) < new Date()) {
      resp = await object.startDownload(name, { acceptDownloadMethods: { getUrl: true } });
      responseUsed = false;
    }

    try {
      responseUsed = true;
      const src = got.stream(resp.url, { retry: false });
      src.on('response', res => {
        contentType = res.headers['content-type'] || contentType;
      });
      const dest = await streamFactory();
      hashStream = new HashStream();
      await pipeline(src, hashStream, dest);

      return;
    } catch (err) {
      // treat non-500 HTTP responses as fatal errors, and retry everything else
      if (err instanceof got.HTTPError && err.response.statusCode < 500) {
        throw err;
      }
      return retriableError(err);
    }
  });

  // now that the download is complete, check the hashes.  Note that a hash
  // verification failure does not result in a retry.
  const observedHashes = hashStream.hashes();
  verifyHashes(observedHashes, resp.hashes);

  return contentType;
};

// verify that all known hashes match, and that at least one of them is an
// "acceptable" hash algorithm.  Throws an exception on verification failure.
const verifyHashes = (observedHashes, expectedHashes) => {
  let someValidAcceptableHash = false;
  for (let algo of Object.keys(expectedHashes)) {
    const computed = observedHashes[algo];
    if (!computed) {
      // ignore unknown hash algorithms
      continue;
    }
    if (computed !== expectedHashes[algo]) {
      throw new Error(`Computed ${algo} hash does not match that from object service`);
    }

    if (ACCEPTABLE_HASHES.has(algo)) {
      someValidAcceptableHash = true;
    }
  }

  if (!someValidAcceptableHash) {
    throw new Error("No acceptable hash algorithm found");
  }
};

export const download = async ({ name, object, streamFactory, retries, delayFactor,
  randomizationFactor, maxDelay }) => {
  const retryCfg = makeRetryCfg({ retries, delayFactor, randomizationFactor, maxDelay });

  const acceptDownloadMethods = {
    getUrl: true,
  };

  const resp = await object.startDownload(name, { acceptDownloadMethods });

  if (resp.method === 'getUrl') {
    return await getUrl({ object, name, resp, streamFactory, retryCfg });
  } else {
    throw new Error("Could not negotiate a download method");
  }
};

export const downloadArtifact = async ({
  taskId, runId, name, queue, streamFactory, retries, delayFactor, randomizationFactor, maxDelay,
}) => {
  const retryCfg = makeRetryCfg({ retries, delayFactor, randomizationFactor, maxDelay });

  let artifact = await (runId === undefined ? queue.latestArtifact(taskId, name) : queue.artifact(taskId, runId, name));

  switch (artifact.storageType) {
    case "reference":
    case "s3": {
      return await s3({ url: artifact.url, streamFactory, retryCfg });
    }

    case "object": {
      const object = new taskcluster.Object({
        rootUrl: queue._options._trueRootUrl,
        credentials: artifact.credentials,
      });
      return await download({ name: artifact.name, object, streamFactory, ...retryCfg });
    }

    case "error": {
      const err = new Error(artifact.message);
      err.reason = artifact.reason;
      throw err;
    }

    default:
      throw new Error(`Unsupported artifact storageType '${artifact.storageType}'`);
  }
};

export default { download, downloadArtifact };

File: clients/client-go/tcobject/download.go
package tcobject

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"

	"github.com/cenkalti/backoff/v3"
	"github.com/orcaman/writerseeker"
	"github.com/taskcluster/httpbackoff/v3"
	"github.com/taskcluster/taskcluster/v96/clients/client-go/internal"
)

type HTTPRetryError = internal.HTTPRetryError

// DownloadToBuf is a convenience method to download an object to an in-memory
// byte slice. Returns the object itself, the Content-Type and Content-Length of
// the downloaded object.
func (object *Object) DownloadToBuf(name string) (buf []byte, contentType string, contentLength int64, err error) {
	writeSeeker := &writerseeker.WriterSeeker{}
	contentType, contentLength, err = object.DownloadToWriteSeeker(name, writeSeeker)
	if err != nil {
		return
	}
	reader := writeSeeker.BytesReader()
	buf = make([]byte, reader.Len())
	_, err = reader.Read(buf)
	return
}

// DownloadToFile is a convenience method to download an object to a file. The
// file is overwritten if it already exists. Returns the Content-Type and
// Content-Length of the downloaded object.
func (object *Object) DownloadToFile(name string, filepath string) (contentType string, contentLength int64, err error) {
	writeSeeker, err := os.Create(filepath)
	if err != nil {
		return "", 0, err
	}
	defer func() {
		err2 := writeSeeker.Close()
		if err == nil {
			err = err2
		}
	}()
	return object.DownloadToWriteSeeker(name, writeSeeker)
}

// getUrlDownload implements the getUrl download method.
func (object *Object) getUrlDownload(rawResponse *DownloadObjectResponse, name string, writeSeeker io.WriteSeeker) (contentType string, contentLength int64, err error) {
	downloadResponse := GetURLDownloadResponse{}
	responseUsed := false
	err = json.Unmarshal(*rawResponse, &downloadResponse)
	if err != nil {
		return
	}

	// wrap the writeSeeker so that we can get the hashes of the received data
	hashingWriter := newHashingWriteSeeker(writeSeeker)

	retryFunc := func() (resp *http.Response, tempError error, permError error) {
		// Explicitly seek to start here, rather than only after a temp error,
		// since not all temporary errors are caught by this code (e.g. status
		// codes 500-599 are handled by httpbackoff library implicitly).
		_, permError = hashingWriter.Seek(0, io.SeekStart)
		if permError != nil {
			// not being able to seek to start is a problem that is unlikely to
			// be solved with retries with exponential backoff, so give up
			// straight away
			return
		}

		// if the URL has expired, fetch a new one, verifying that each response
		// is used at least once to avoid infinitely re-fetching
		if responseUsed && time.Time(downloadResponse.Expires).Before(time.Now()) {
			rawResponse, err = object.StartDownload(
				name,
				&DownloadObjectRequest{
					AcceptDownloadMethods: SupportedDownloadMethods{GetURL: true},
				},
			)
			if err != nil {
				return
			}
			err = json.Unmarshal(*rawResponse, &downloadResponse)
			if err != nil {
				return
			}
			if downloadResponse.Method != "getUrl" {
				err = fmt.Errorf("got unexpected download method %v", downloadResponse.Method)
				return
			}
			responseUsed = false
		}

		// Get the URL.  Note that this adds `Acccept-Encoding: gzip` and will
		// automatically un-gzip a response if necessary.
		responseUsed = true
		resp, tempError = http.Get(downloadResponse.URL)

		// httpbackoff handles http status codes, so we can consider all errors worth retrying here
		if tempError != nil || resp.StatusCode != http.StatusOK {
			// temporary error!
			return
		}
		defer resp.Body.Close()
		contentType = resp.Header.Get("Content-Type")
		contentLength, tempError = io.Copy(hashingWriter, resp.Body)
		return
	}
	var resp *http.Response
	// HTTP status codes handled here automatically
	client := object.HTTPBackoffClient
	if client == nil {
		client = &httpbackoff.Client{
			BackOffSettings: backoff.NewExponentialBackOff(),
		}
	}
	var attempts int
	resp, attempts, err = client.Retry(retryFunc)
	if err != nil {
		err = HTTPRetryError{
			Attempts: attempts,
			Err:      err,
		}
	}
	resp.Body.Close()
	if err != nil {
		return
	}

	// verify hashes if no error has occurred so far.  Note that the download is not
	// retried if hash verification fails.
	if err == nil {
		var hashes map[string]string
		hashes, err = unmarshalHashes(downloadResponse.Hashes)
		if err != nil {
			err = HTTPRetryError{
				Attempts: attempts,
				Err:      err,
			}
			return
		}
		err = verifyHashes(hashingWriter, hashes)
		if err != nil {
			err = HTTPRetryError{
				Attempts: attempts,
				Err:      err,
			}
			return
		}
	}

	// if there is an error, empty everything else out so that callers do not use the
	// untrusted returned data
	if err != nil {
		contentType = ""
		contentLength = 0
		_, _ = hashingWriter.Seek(0, io.SeekStart)
		return
	}
	return
}

// verifyHashes verifies that the hashes observed by the hashingWriter match
// those in hashes, where present, and that at least one is present.
func verifyHashes(hashingWriter *hashingWriteSeeker, expectedHashes map[string]string) error {
	observedHashes, err := hashingWriter.hashes()
	if err != nil {
		return err
	}

	// this will be set to true when an acceptable (that is, not deprecated) hash
	// algorithm is found with a valid hash.
	var someValidAcceptableHash bool

	for algo, oh := range observedHashes {
		if eh, ok := expectedHashes[algo]; ok {
			if oh != eh {
				return fmt.Errorf("validation of object data's %s hash failed", algo)
			}
			for _, a := range ACCEPTABLE_HASHES {
				if algo == a {
					someValidAcceptableHash = true
				}
			}
		}
	}

	if !someValidAcceptableHash {
		return errors.New("no acceptable hashes found in object metadata")
	}

	return nil
}

// DownloadToWriteSeeker downloads the named object from the object service and
// writes it to writeSeeker, retrying if intermittent errors occur. Returns
// the Content-Type and Content-Length of the downloaded object.
func (object *Object) DownloadToWriteSeeker(name string, writeSeeker io.WriteSeeker) (contentType string, contentLength int64, err error) {
	downloadObjectResponse, err := object.StartDownload(
		name,
		&DownloadObjectRequest{
			AcceptDownloadMethods: SupportedDownloadMethods{
				GetURL: true,
			},
		},
	)
	if err != nil {
		return "", 0, err
	}

	var bareResp struct{ Method string }
	err = json.Unmarshal(*downloadObjectResponse, &bareResp)
	if err != nil {
		return "", 0, err
	}

	switch bareResp.Method {
	case "getUrl":
		return object.getUrlDownload(downloadObjectResponse, name, writeSeeker)
	}
	return "", 0, fmt.Errorf("unknown download method %q for object %q", bareResp.Method, name)
}
